langchain/utilities/wikipedia.py

```diff
@@ -41,23 +47,70 @@ def validate_environment(cls, values: Dict) -> Dict:
 
     def run(self, query: str) -> str:
         """Run Wikipedia search and get page summaries."""
-        search_results = self.wiki_client.search(query[:WIKIPEDIA_MAX_QUERY_LENGTH])
+        page_titles = self.wiki_client.search(query[:WIKIPEDIA_MAX_QUERY_LENGTH])
         summaries = []
-        len_search_results = len(search_results)
-        if len_search_results == 0:
+        for page_title in page_titles[: self.top_k_results]:
+            if wiki_page := self._fetch_page(page_title):
+                if summary := self._formatted_page_summary(page_title, wiki_page):
+                    summaries.append(summary)
+        if not summaries:
             return "No good Wikipedia Search Result was found"
-        for i in range(min(self.top_k_results, len_search_results)):
-            summary = self.fetch_formatted_page_summary(search_results[i])
-            if summary is not None:
-                summaries.append(summary)
         return "\n\n".join(summaries)
 
-    def fetch_formatted_page_summary(self, page: str) -> Optional[str]:
+    @staticmethod
+    def _formatted_page_summary(page_title: str, wiki_page: Any) -> Optional[str]:
+        return f"Page: {page_title}\nSummary: {wiki_page.summary}"
+
+    def _page_to_document(self, page_title: str, wiki_page: Any) -> Document:
+        main_meta = {
+            "title": page_title,
+            "summary": wiki_page.summary,
+        }
+        add_meta = (
+            {
+                "categories": wiki_page.categories,
+                # "coordinates": wiki_page.coordinates,
+                "page_url": wiki_page.url,
+                "image_urls": wiki_page.images,
+                "related_titles": wiki_page.links,
+                "parent_id": wiki_page.parent_id,
+                "references": wiki_page.references,
+                "revision_id": wiki_page.revision_id,
+                "sections": wiki_page.sections,
+            }
+            if self.load_all_available_meta
+            else {}
+        )
+        doc = Document(
+            page_content=wiki_page.content,
+            metadata={
+                **main_meta,
+                **add_meta,
+            },
+        )
+        return doc
+
+    def _fetch_page(self, page: str) -> Optional[str]:
         try:
-            wiki_page = self.wiki_client.page(title=page, auto_suggest=False)
-            return f"Page: {page}\nSummary: {wiki_page.summary}"
+            return self.wiki_client.page(title=page, auto_suggest=False)
         except (
             self.wiki_client.exceptions.PageError,
             self.wiki_client.exceptions.DisambiguationError,
         ):
             return None
+
+    def load(self, query: str) -> List[Document]:
```

Comment:
tangent: @hwchase17 would make for a cool Retriever as well, maybe worth coming up with a little demo (in a sep pr)